[WIP] Add gzip decoding by default by inspecting header.

echochamber opened the pull request:
Some discussion already [here](https://github.com/seanmonstar/reqwest/issues/3)

seanmonstar commented on src/client.rs:
```diff
@@ -256,24 +253,54 @@ impl<'a> RequestBuilder<'a> {
 impl Response {
+    /// Constructs a response from a hyper request
+    /// Uses the correct variant by inspecting the content-encoding header.
+    pub fn from_hyper_response(res: ::hyper::client::Response) -> Self {
```
I'd remove the `pub` on this function, since it's only used internally.

seanmonstar commented on src/client.rs:
```diff
@@ -256,24 +253,54 @@ impl<'a> RequestBuilder<'a> {
+    /// Uses the correct variant by inspecting the content-encoding header.
+    pub fn from_hyper_response(res: ::hyper::client::Response) -> Self {
+        let mut is_gzip = false;
+        match res.headers.get::<ContentEncoding>() {
```
As mentioned before, we should be checking `TransferEncoding`, not `ContentEncoding`.

seanmonstar commented on src/client.rs:
```diff
@@ -256,24 +253,54 @@ impl<'a> RequestBuilder<'a> {
 #[derive(Debug)]
-pub struct Response {
-    inner: ::hyper::client::Response,
+pub enum Response {
```
Let's keep the internals private; enum variants are all public. So instead, something like:

```rust
pub struct Response {
    inner: Decoder,
}

enum Decoder {
    Plaintext(::hyper::client::Response),
    Gzip(..),
}
```

echochamber commented:
Hi @seanmonstar.

So I'm kind of in a weird spot where I think I that I've gotten the code here to a working state (or close to working) but I'm having difficulties testing it. Before I explain the issues, I just want to clarify that I'm pretty far out of my normal comfort zone with anything lower level than scripting languages. Lets say I have a rough understanding of basic lower level programming concepts in theory but have very little applied practice. So appologies in advance if any of these mistakes/issues I'm running into are super trivial.

I thought the easiest way to test this was just find a site that used `Transfer-Encoding: gzip` but had a hard time actually finding one. So in order to manually test the actual gzip decoding was working correctly, I made some tweaks that I was planning to undo once I verified that it was working and updated my PR to address your comments. These changes are in the most recent commit in this PR (again, just for reference, I will undo all the changes in that commit once I've verified its working correctly).

#### My changes purely for testing this are:

* I changed the header the gzip decoding step was checking back to ContentEncoding (from TransferEncoding) because I noticed http://google.com would use `ContentEncoding: gzip` if my request included the `Accept-Encoding: gzip` header.
* I made another example that included that header, and printed the output same as the simple example. If all was working correctly, then that example should print out the content same as normal.

The issue I ended up having is a bit strange, and I'm not really certain how to proceed debugging it. What appears to be happening is the gzip decoder is successfully decoding some part of the stream, but then is panicing because its missing the gzip header?

#### My best guess for the source of the issue:

The response is sending the content back in chunks, and the gzip header is only included as part of the first chunk, so when the stream receives the second chunk and its lacking the header, it panics. However, I am very lost on how I would go about verifying this guess or how to proceed debugging from here.

Again, apologies if I'm missing something trivial due to my experience and thanks in advance for any help you can provide!

seanmonstar commented on src/client.rs:
```diff
@@ -282,12 +325,27 @@ impl Response {
+                hyper_response.read(buf)
+            },
+            &mut Decoder::Gzip(ref mut hyper_response) => {
+                GzDecoder::new(BufReader::new(hyper_response))?.read(buf)
```
This `GzDecoder` should be constructed in `from_hyper_response`, and saved. The error you are seeing is that part of the stream is being decoded, and then the decoder is being thrown away. The `BufReader` shouldn't be necessary, because the `hyper::client::Response` already have a `BufReader` inside it.

echochamber replied:
Question: So if I am storing the ::hyper::client::Response object insider the decoder, I don't believe there is a way for me to access it because constructor the decoder takes ownership of it. But since `::reqwest::client::Response` still needs to support the `status()`, `headers()`, and `version()` methods we still need access to that data in some form. As I see it currently, we can change the definitions of the `Decoder::Gzip` variant to be something like:

```rust
enum Decoder {
    
    PlainText(::hyper::client::Response),
    
    Gzip {
        decoder: ::flate2::read::GzDecoder<::hyper::client::Response>,
        headers: ::hyper::headers::Headers,
        version: ::hyper::version::HttpVersion,
        status: ::hyper::status::StatusCode,
    }
}
```

and then we just do something like this where we copy the data before the request object ownership is lost.

```rust
    fn from_hyper_response(res: ::hyper::client::Response) -> Self {

        // ...

        if is_gzip {
            return Decoder::Gzip {
                status: res.status.clone(),
                version: res.version.clone(),
                headers: res.headers.clone(),
                decoder: GzDecoder::new(res).unwrap())
            };
        } else {
            return Decoder::PlainText(res);
        }
    }
```

but this strikes me as a poor solution. Storing fields named status, version, headers, on type named Decoder::Gzip seems incorrect to me. I actually think that data might be better off stored directly on the `::reqwest::client::Request` object, and then the inner `::hyper::client::Request` isn't stored inside the reqwest struct at all, instead maybe we just keep handle to the stream or a Read object of some sort.

seanmonstar commented:
Getting closer! I'll be working on adding a way to easily add integration tests, and that way should be pretty simple to define a test case where a server sends back a gzip header, and hard code in the gzipped body, so we can know for sure that it should always decode to an exact value.

echochamber commented:
@seanmonstar Ok, it works now! I tested it and everything 🎉. Although I'm sure there's plenty of ways to improve the code I've written and make it more idiomatic and all that fun stuff. My comment here https://github.com/seanmonstar/reqwest/pull/25#issuecomment-262437229 is still relevant. 

And then there's still whether we should check the `TransferEncoding` header or `ContentEncoding` and automatically decode it. I thought since this is a library who's goal is convenience, it would be best to look at the some http client libraries in scripting languages to see how they do automatic encoding/decoding:

## Popular http client libs from scripting languages

### PHP Guzzle autodecoding: 

Does it based on how the client object is configured. But it checks the Content-Encoding header.

https://github.com/guzzle/guzzle/blob/113071af3b2b9eb96b05dab05472cbaed64a3df4/src/Handler/StreamHandler.php#L158-L166

### Ruby httpclient auto gzip decoding:

Same as guzzle, does it based on how its configured but also checks Content-Encoding.

https://github.com/nahi/httpclient/blob/bcc2c0a52aeda3a96025e33ae65b0bdf7f708331/lib/httpclient/session.rb#L570-L575

### NodeJS request/request:

Seems like this one will always auto decompress gziped if a response has Content-Encoding: gzip. No config option available
https://github.com/request/request/blob/9f702bf4b5d075035c28eea55dda44d3a0235926/request.js#L945-L951

### Python requests

Same as with the PHP/Ruby/NodeJS libs, does it based on configuration.

https://github.com/kennethreitz/requests/blob/362da46e9a46da6e86e1907f03014384ab210151/requests/packages/urllib3/response.py#L87-L90

seanmonstar commented:
Excellent work! I've added integration tests, so this PR could include one to make sure it works. It'd mean adding something like the following to `tests/client.rs`:

```rust
#[test]
fn test_gzip_response() {
    let server = server! {
        request: expected_request_bytes,
        response: b"\
            HTTP/1.1 200 OK\r\n\
            Content-Encoding: gzip\r\n\
            Content-Length: $gzipped_body_length\r\n\
            \r\n\
            $gzipped_body_of_something_small\
            ",
    };
    let res = reqwest::get(&format!("http://{}/gzip")).unwrap();
    // create a String, and read_to_string, and assert that the string equals the unzipped body.
}
```

Besides the the test, the last thing I'm thinking is probably to add an option to `Client` so people can enable/disable auto gzip decoding. `client.gzip(false)` or `client.set_gzip(false)` or similar.

echochamber commented:
@seanmonstar Awesome, big thanks for writing that. I'll see if I can have this PR good to go in the next day or so.

Separately, I'll open an issue for adding a paragraph or so rustdocs to the new `server!()` macro. I can save you the trouble of writing them since its pretty simple.

seanmonstar commented:
Hey @echochamber, were you still interested in pushing on this?

echochamber commented:
@seanmonstar Oh sorry, I forgot I still had this open. Let me see how difficult it is to resolve the merge conflicts over the weekend.

echochamber commented:
@seanmonstar Sorry, wasn't feeling 100% over the weekend. Jumping back into this PR tonight/tomorrow. Thanks for the patience =)

echochamber commented:
@seanmonstar Going to close this PR. Haven't had much energy for writing code after work the last few weeks. If I pick this issue up again, I'll open a new PR for it. Sorry.
